apps/aeoracle/src/aeo_state_tree.erl

```diff
@@ -0,0 +1,166 @@
+%%%===================================================================
+%%% Types
+%%%===================================================================
+
+-type mtree() :: aeu_mtrees:tree(binary(), binary()).
+-type interaction() :: aeo_interaction:interaction().
+-type oracle() :: aeo_oracles:oracle().
+-type cache() :: gb_sets:set({integer(), binary()}).
+-type block_height() :: non_neg_integer().
+
+-record(oracle_tree, { mtree = gb_merkle_trees:empty() :: mtree()
+                     , cache = cache_new() :: cache()
+                     }).
+
+-opaque tree() :: #oracle_tree{}.
+
+-export_type([ tree/0
+             ]).
+
+%%%===================================================================
+%%% API
+%%%===================================================================
+
+-spec empty() -> tree().
+empty() ->
+    MTree = aeu_mtrees:empty(),
+    #oracle_tree{ mtree = MTree
+                , cache = cache_new()
+                }.
+
+-spec prune(block_height(), tree()) -> tree().
+prune(Height, #oracle_tree{} = Tree) ->
+    %% TODO: We need to know what we pruned as well
+    %% Oracle information should be around for the expiry block
+    %% since we prune before the block, use Height - 1 for pruning.
+    int_prune(Height - 1, Tree).
+
+-spec enter_interaction(interaction(), tree()) -> tree().
+enter_interaction(I, Tree) ->
+    Id = aeo_interaction:id(I),
+    Serialized = aeo_interaction:serialize(I),
+    Expires = aeo_interaction:expires(I),
+    enter_common(Id, Expires, Serialized, Tree).
+
+-spec get_interaction(binary(), tree()) -> interaction().
+get_interaction(Id, Tree) ->
+    aeo_interaction:deserialize(aeu_mtrees:get(Id, Tree#oracle_tree.mtree)).
+
+-spec lookup_interaction(binary(), tree()) -> {'value', interaction()} | none.
+lookup_interaction(Id, Tree) ->
+    case aeu_mtrees:lookup(Id, Tree#oracle_tree.mtree) of
+      {value, Val} -> {value, aeo_interaction:deserialize(Val)};
+      none -> none
+    end.
+
+-spec enter_oracle(oracle(), tree()) -> tree().
+enter_oracle(O, Tree) ->
+    Id = aeo_oracles:id(O),
+    Serialized = aeo_oracles:serialize(O),
+    Expires = aeo_oracles:expires(O),
+    enter_common(Id, Expires, Serialized, Tree).
+
+-spec get_oracle(binary(), tree()) -> oracle().
+get_oracle(Id, Tree) ->
+    aeo_oracles:deserialize(aeu_mtrees:get(Id, Tree#oracle_tree.mtree)).
+
+-spec lookup_oracle(binary(), tree()) -> {'value', oracle()} | 'none'.
+lookup_oracle(Id, Tree) ->
+    case aeu_mtrees:lookup(Id, Tree#oracle_tree.mtree) of
+        {value, Val}  -> {value, aeo_oracles:deserialize(Val)};
+        none -> none
+    end.
+
+-spec root_hash(tree()) -> {ok, aeu_mtrees:root_hash()} | {error, empty}.
+root_hash(#oracle_tree{mtree = MTree}) ->
+    aeu_mtrees:root_hash(MTree).
+
+-spec oracle_list(tree()) -> list(oracle()).
+oracle_list(#oracle_tree{mtree = MTree}) ->
+    ToOracle = fun(MaybeO) ->
+                   try [aeo_oracles:deserialize(MaybeO)]
+                   catch _:_ -> [] end end,
```

Comment:
I am concerned about this catch. Should not invalid oracles not slip in the trees in the first place?

Reply:
I see - this is test only.

Reply:
This is a function only exported for test. (-ifdef(TEST).)

The try catch is there to only get oracles (and not interactions)